src/lib.rs

```diff
@@ -546,27 +581,43 @@ pub unsafe trait FromBytes: FromZeroes {
 ///
 /// # Safety
 ///
-/// If `T: AsBytes`, then unsafe code may assume that it is sound to treat any
-/// instance of the type as an immutable `[u8]` of length `size_of::<T>()`. If a
-/// type is marked as `AsBytes` which violates this contract, it may cause
+/// *This section describes what is required in order for `T: AsBytes`, and what
+/// unsafe code may assume of such types. `#[derive(AsBytes)]` only permits
+/// types which satisfy these requirements. If you don't plan on implementing
+/// `AsBytes` manually, and you don't plan on writing unsafe code that operates
+/// on `AsBytes` types, then you don't need to read this section.*
+///
+/// If `T: AsBytes`, then unsafe code may assume that:
+/// - ...it is sound to treat any `t: T` as an immutable `[u8]` of length
+///   `size_of_val(t)`.
+/// - ...given `t: &T`, it is sound to construct a `b: &[u8]` where `b.len() ==
+///   size_of_val(t)` at the same address as `t`, and it is sound for both `b`
+///   and `t` to be live at the same time.
+///
+/// If a type is marked as `AsBytes` which violates this contract, it may cause
 /// undefined behavior.
 ///
-/// If a type has the following properties, then it is safe to implement
-/// `AsBytes` for that type
+/// If a type has the following properties, then it is sound to implement
+/// `AsBytes` for that type:
```

Comment:
So, there are multiple kinds of safety requirements listed here.

First it states what operations that unsafe code are allowed to perform. Then, it states that certain types are sound to implement for.

However, there's a missing step here. It should first say what _core properties_ a type must uphold in order for those operations to be sound. Those core safety requirements are:

- No bytes that can possibly be uninitialized:
  - No padding
  - No `MaybeUninit`
- No `UnsafeCell`

These are the requirements for every data type, struct, enum, union, etc.

Then, it can expand upon what types naturally fit these requirements.
The bits about "if the type is a struct" are essentially _requirements for the derive_ to ensure that the above safety requirements are met, and are stricter than the actual safety requirements. Some of them are also requirements to guarantee _stability_ of sound implementation, such as requiring `#[repr(C)]` on structs.

This is less applicable for `FromBytes` since the core safety requirements are essentially identical to the guarantees unsafe code can assume. Might be worth throwing in a "it cannot have any invalid representations"?

Reply:
Filed https://github.com/google/zerocopy/issues/288 to track this.